File: 프로그래머스/2/42584. 주식가격/주식가격.cpp
#include <string>
#include <vector>

using namespace std;

vector<int> solution(vector<int> prices) {
    
    vector<int> result;
    
    // 이중 반복문 구현
    for(int i = 0 ; i < prices.size() ; i++) {
        int sec = 0; // 경과 시점
        for(int j = i + 1 ; j < prices.size() ; j++) {
            // 가격이 떨어지지 않았을 경우
            if(prices[i] <= prices[j]) {
                // 경과 시점 증가
                sec++;
            }
            else {
                sec++;
                break;
            }
        }
        result.push_back(sec);
    }
    
    return result;
}

File: 프로그래머스/2/42584. 주식가격/주식가격.java
import java.util.*;

class Solution {
    public int[] solution(int[] prices) {
        
        
        Stack<int[]> stack = new Stack<>();
        
        int[] answer = new int[prices.length];
        int time = 0;
        
        for(int i = 0 ; i < prices.length ; i++){
            time++;
            
            // 현재
            int price = prices[i];
            // 감소한 거 뺴기
            while(!stack.isEmpty() && prices[stack.peek()[0]] > price){
                int[] s = stack.pop();
                answer[s[0]] = time - s[1];
            }
            
            stack.push(new int[]{i, time});
            
        }
        
        // 남아있는 거 정리
        while(!stack.isEmpty()){
            int[] s = stack.pop();
            answer[s[0]] = time - s[1];
        }
        
        return answer;
    }
}